src/master/matoclserv.cc

```diff
@@ -1946,68 +1909,38 @@ void matoclserv_fuse_reserved_inodes(matoclserventry *eptr,const uint8_t *data,u
 	}
 
 	ptr = data;
-	ofpptr = &(eptr->sesdata->openedfiles);
 	length >>= 2;
-	if (length) {
+	std::set<uint32_t> inodes_to_reserve;
+	// read in advance all the files to reserve
+	while (length) {
 		length--;
-		inode = get32bit(&ptr);
-	} else {
-		inode=0;
+		inodes_to_reserve.insert(get32bit(&ptr));
 	}
 
 	FsContext context = FsContext::getForMaster(eventloop_time());
 	changelog_disable_flush();
-	while ((ofptr=*ofpptr) && inode>0) {
-		if (ofptr->inode<inode) {
-			fs_release(context, ofptr->inode, eptr->sesdata->sessionid);
-			*ofpptr = ofptr->next;
-			free(ofptr);
-		} else if (ofptr->inode>inode) {
-			if (fs_acquire(context, inode, eptr->sesdata->sessionid) == SAUNAFS_STATUS_OK) {
-				ofptr = (filelist*)malloc(sizeof(filelist));
-				passert(ofptr);
-				ofptr->next = *ofpptr;
-				ofptr->inode = inode;
-				*ofpptr = ofptr;
-				ofpptr = &(ofptr->next);
-			}
-			if (length) {
-				length--;
-				inode = get32bit(&ptr);
-			} else {
-				inode=0;
-			}
+	auto it = eptr->sesdata->openedfiles.begin();
+	while (it != eptr->sesdata->openedfiles.end()) {
+		uint32_t openFileIno = *it;
+		if (inodes_to_reserve.count(openFileIno) == 0) {
+			// erase files not belonging to the reserve inodes list provided
+			fs_release(context, openFileIno, eptr->sesdata->sessionid);
+			it = eptr->sesdata->openedfiles.erase(it);
 		} else {
-			ofpptr = &(ofptr->next);
-			if (length) {
-				length--;
-				inode = get32bit(&ptr);
-			} else {
-				inode=0;
-			}
+			// skip files already in session
+			it++;
+			// no need to remind this file as reserved, as it is already open
+			inodes_to_reserve.erase(openFileIno);
 		}
 	}
-	while (inode>0) {
-		if (fs_acquire(context, inode, eptr->sesdata->sessionid) == SAUNAFS_STATUS_OK) {
-			ofptr = (filelist*)malloc(sizeof(filelist));
-			passert(ofptr);
-			ofptr->next = *ofpptr;
-			ofptr->inode = inode;
-			*ofpptr = ofptr;
-			ofpptr = &(ofptr->next);
-		}
-		if (length) {
-			length--;
-			inode = get32bit(&ptr);
-		} else {
-			inode=0;
+
+	for (const auto &inode_to_reserve : inodes_to_reserve) {
+		if (fs_acquire(context, inode_to_reserve, eptr->sesdata->sessionid) ==
+		    SAUNAFS_STATUS_OK) {
+			// insert the ones that are not already open missing
```

Comment:
This comment can be improved. Maybe the `missing` at the end is not needed?
```cpp
// Insert successfully reserved inodes into the opened files set
```

Reply:
Applied.